### PR TITLE
Allow import of multiple programs per address location

### PR DESCRIPTION
**Pull request by turbolent**
Work towards #221

Refactor how programs are imported.

Previously, the code that handles imports assumed that each import location (e.g. an address like `0x1` maps to one program, and the specifically given identifiers in the import statement (if any, otherwise all identifiers) are imported from this single program.

The code that handles imports was previously partially implemented in the AST package and partially in the checker package (`sema`). This PR removes the code in the AST package and implements it fully in the checker package. It is now split into two phases: 

1. Resolution of the import statement. 

    The default case is that one location is resolved to one location (itself), though now an address location may also be resolved into multiple "address contract" locations (naming suggestions welcome).

    For example, an import statement `import a, b from 0x1` specifies the specific identifiers `a` and `b` and the address location `0x1`. It might be resolved into just the location itself, i.e. the address location `0x1`, but can now also resolved into multiple locations, e.g. the address contract location `0x1.a` and the address contract location `0x1.b`.

2. Acquiring the programs for the resolved imports. For each resolved import a separate program can be returned.

The split of the import code into resolution and acquisition required changes to `runtime.Interface`.
Exposing the two phases allows greater flexibility and implementing backwards compatibility for existing deployed contracts in the host environment. 

If we want to keep backwards compatibility of the interface we can revert the changes and hard-code how addresses are resolved, however, I can't quite see how we could keep backwards compatibility for already deployed contracts.

In addition, this refactor also enables repeated imports from the same locations, which was previously not supported, e.g. `import a from 0x1; import b from 0x1`.

While I adjusted the code for the REPL, I added support for importing local files.

### CODE REVIEW

File: runtime/ast/program.go [hunk 1]
```diff
@@ -104,95 +102,6 @@ func (p *Program) TransactionDeclarations() []*TransactionDeclaration {
 	return p._transactionDeclarations
 }
 
-// ImportedPrograms returns the sub-programs imported by this program, indexed by location ID.
```
Comment (turbolent) [inline runtime/ast/program.go:107]: Remove all code related to imports from the AST package

File: runtime/cmd/cmd.go [hunk 2]
```diff
@@ -84,39 +93,23 @@ func mustClosure(filename string, codes map[string]string) func(error) {
...
-	codes := map[string]string{}
-
+func PrepareProgram(code string, filename string, codes map[string]string) (*ast.Program, func(error)) {
 	must := mustClosure(filename, codes)
 
 	program, err := parser2.ParseProgram(code)
 	codes[filename] = code
 	must(err)
 
-	err = program.ResolveImports(func(location ast.Location) (program *ast.Program, err error) {
```
Comment (turbolent) [inline runtime/cmd/cmd.go:105]: Now implemented as a checker import handler: https://github.com/onflow/cadence/pull/333/files#diff-040f568b71e2279bc7135a9b78a251b4R134

File: runtime/cmd/execute/repl.go [hunk 3]
```diff
@@ -53,6 +57,37 @@ func RunREPL() {
 
 			println(colorizeResult(value))
 		},
+		[]sema.Option{
+			sema.WithImportHandler(
```
Comment (turbolent) [inline runtime/cmd/execute/repl.go:61]: Add support for imports of files (string locations) in the REPL

File: runtime/interface.go [hunk 4]
```diff
@@ -36,9 +37,13 @@ type Block struct {
...
 
+type ResolvedLocation = sema.ResolvedLocation
+
 type Interface interface {
-	// ResolveImport resolves an import of a program.
-	ResolveImport(Location) ([]byte, error)
+	// ResolveLocation resolves an import location.
+	ResolveLocation(identifiers []Identifier, location Location) []ResolvedLocation
+	// GetCode returns the code at a given location
+	GetCode(location Location) ([]byte, error)
```
Comment (turbolent) [inline runtime/interface.go:46]: Split the single import function into two function: resolving the import location, and getting the code for a resolved location

File: runtime/tests/interpreter/import_test.go [hunk 5]
```diff
@@ -115,3 +118,144 @@ func TestInterpretVirtualImport(t *testing.T) {
 		value,
 	)
 }
+
+func TestInterpretImportMultipleProgramsFromLocation(t *testing.T) {
```
Comment (turbolent) [inline runtime/tests/interpreter/import_test.go:122]: This is the most important new test: It demonstrates how two declarations (`a` and `b`) can be imported from the same location (address location `0x1`), are resolved to two locations, which are backed by two programs.

File: runtime/cmd/cmd.go [hunk 6]
```diff
@@ -131,13 +124,42 @@ var typeDeclarations = append(
...
 // and a filename which is used for pretty-printing errors, if any
-func PrepareChecker(program *ast.Program, filename string, must func(error)) (*sema.Checker, func(error)) {
+func PrepareChecker(program *ast.Program, filename string, codes map[string]string, must func(error)) (*sema.Checker, func(error)) {
 	location := runtime.FileLocation(filename)
 	checker, err := sema.NewChecker(
 		program,
 		location,
 		sema.WithPredeclaredValues(valueDeclarations.ToValueDeclarations()),
 		sema.WithPredeclaredTypes(typeDeclarations),
+		sema.WithImportHandler(
```
Comment (Kay-Zee) [inline runtime/cmd/cmd.go:134]: Nice!

File: runtime/cmd/cmd.go [hunk 7]
```diff
@@ -131,13 +124,42 @@ var typeDeclarations = append(
...
+							fmt.Errorf("cannot import `%s`. only files are supported", location),
+						},
+					}
+				}
+
+				importChecker, err := checker.EnsureLoaded(
+					location,
+					func() *ast.Program {
+						filename := string(stringLocation)
+						imported, _ := PrepareProgramFromFile(filename, codes)
```
Comment (Kay-Zee) [inline runtime/cmd/cmd.go:150]: Is this an err that's skipped?

File: runtime/runtime_test.go [hunk 8]
```diff
@@ -2337,48 +2350,6 @@ func TestRuntimeTransaction_AddPublicKey(t *testing.T) {
 	}
 }
 
-func TestRuntimeCyclicImport(t *testing.T) {
```
Comment (Kay-Zee) [inline runtime/runtime_test.go:2340]: Is this moved somewhere? Or covered by another test? `TestInterpretImportMultipleProgramsFromLocation` maybe?

File: runtime/tests/interpreter/import_test.go [hunk 9]
```diff
@@ -115,3 +118,150 @@ func TestInterpretVirtualImport(t *testing.T) {
...
+
+	addressLocation := ast.AddressLocation{0x1}
+
+	importedCheckerA, err := checker.ParseAndCheckWithOptions(t,
+		`
+          pub fun a(): Int {
+              return 1
+          }
+
+          pub fun b(): Int {
```
Comment (Kay-Zee) [inline runtime/tests/interpreter/import_test.go:140]: In this scenario, there'd be no way to access this `b()` func?

**Review by Kay-Zee — APPROVED** [review]
Sorry this took so long, there was a lot of content to go through.

I spent a majority of my focus on the tests to make sure they make sense to me. Had a few small questions, but otherwise LGTM

File: runtime/tests/interpreter/import_test.go (hunk 9 above)
Comment (turbolent) [reply runtime/tests/interpreter/import_test.go:140]: Good question!

Correct, this function is only declared to ensure that it is **not** imported in the `import a, b from 0x1` statement.

Likewise, [the function `a` below](https://github.com/onflow/cadence/pull/333/files#diff-f759a5f9b49ae7e692b50eb5926c786cR159) should not be imported either. 

I'll add comments to make this test clearer

File: runtime/cmd/cmd.go (hunk 7 above)
Comment (turbolent) [reply runtime/cmd/cmd.go:150]: It looks like it, but the `PrepareProgramFromFile` function does not return an error as it's second return value, but a `must` function. 

This got introduced at some point and I don't like it, because it's not Go idiomatic and confusing. It would be nice to refactor that at some point.

If you look at `PrepareProgramFromFile`, it already checks the error using the must function, then just returns it so that callers have a convenient way of reporting an error through a single function.

File: runtime/runtime_test.go (hunk 8 above)
Comment (turbolent) [reply runtime/runtime_test.go:2340]: This PR initially removed an explicit check for cyclic imports. Cyclic imports where still not supported, but instead of getting an dedicated error for cyclic imports, the checker would still fail, but only through implicit "not declared" errors.

Thinking about it a bit more, this is not very user-friendly: a developer might only see a "not declared" error, see they have a valid declaration, but would not realize the actual cause is a cyclic import. 

I've added the explicit check and error back in [`2b713fc` (#333)](https://github.com/onflow/cadence/pull/333/commits/2b713fcff712cbb54271811a50f9e15c1828503a).

Regarding this test case in particular: This is a runtime check, but all it really tests is that the checker detects the cyclic import, so it shouldn't be a runtime test in the first place, but rather a checker test.

I've improved https://github.com/onflow/cadence/pull/333/files#diff-4fb398062392664e4c2200a9025f9f42R543 and also added another test for a mutually recursive import here:
https://github.com/onflow/cadence/pull/333/files#diff-4fb398062392664e4c2200a9025f9f42R586.
Basically, these two checker tests cover this test and we don't need the runtime test here.

File: runtime/tests/interpreter/import_test.go (hunk 9 above)
Comment (turbolent) [reply runtime/tests/interpreter/import_test.go:142]: Improved this in [`36421c5` (#333)](https://github.com/onflow/cadence/pull/333/commits/36421c5375832e8f4de5d26fdf78726b7c530d01)
